compiler/optimizer/ValuePropagationCommon.cpp

```diff
@@ -1279,6 +1287,109 @@ void OMR::ValuePropagation::transformArrayCopyCall(TR::Node *node)
+            node->getGlobalIndex(), node, transformTheCall, TR::Compiler->om.areFlattenableValueTypesEnabled(),
+            srcObjNode->getGlobalIndex(), srcObjNode, dstObjNode->getGlobalIndex(), dstObjNode, srcVN, dstVN);
+         }
+
+      static char *disableNullRestrictedArrayCopyXForm = feGetEnv("TR_DisableNullRestrictedArraycopyXForm");
+
+      // JEP 401: If null restricted value type is enabled, we need to preform null check on the value being stored
+      // in order to throw a NullPointerException if the array is null-restricted and the value to write is null.
+      // If it is this case, System.arraycopy cannot be transformed into arraycopy instructions.
+      //
+      if (transformTheCall &&
+          TR::Compiler->om.areFlattenableValueTypesEnabled() &&  // Null restricted value type is enabled
+          !disableNullRestrictedArrayCopyXForm &&
+          !isStringCompressedArrayCopy &&
+          !isStringDecompressedArrayCopy &&
+          !primitiveArray1 &&
+          !primitiveArray2 &&
+          (copyLen != _constantZeroConstraint)) // Not zero length copy
+         {
+         isDstArrayCompTypePrimitiveValueType = isArrayCompTypePrimitiveValueType(dstObject);
+         isSrcArrayCompTypePrimitiveValueType = isArrayCompTypePrimitiveValueType(srcObject);
+
+         switch (isDstArrayCompTypePrimitiveValueType)
+            {
+            case TR_yes:
+               {
+               if (isSrcArrayCompTypePrimitiveValueType == TR_yes)
+                  {
+                  // Array flattening is not enabled
+                  //    - If both source and destination arrays are primitive VT, they don't contain
+                  //      NULL value. There is no need to check null store
+                  //    - Also because flattening is not enabled, we don't need to concern about copying
+                  //      between flattened arrays
+                  //
+                  // Array flattening is enabled
+                  //    - If both source and destination arrays are primitive VT
+                  //        - (1) Both arrays are not flattened, there is no need to do anything
+                  //        - (2) Both arrays are flattened, elementSize needs to be updated in order to
+                  //          use arraycopy instructions
+                  //        - (3) Either of the arrays might or might be flattened, System.arraycopy
+                  //          should not be transformed into arraycopy instructions.
+                  //
+                  if (isValueTypeArrayFlatteningEnabled)
+                     {
+                     if (trace())
+                        traceMsg(comp(), "%s: n%dn %p  isArrayElementFlattened dst %d src %d\n", __FUNCTION__, node->getGlobalIndex(), node,
+                           isArrayElementFlattened(dstObject), isArrayElementFlattened(srcObject));
```

Comment:
>you could just open an issue and mark it as "Help wanted" so that the various places that print values of that type could be cleaned up all at once

This is a good idea! I'll create an issue for this work.

Reply:
Created https://github.com/eclipse-openj9/openj9/issues/17956